### PR TITLE
use ssb-meta-feeds with v1 tree structure

### PR DESCRIPTION
**Pull request by staltz**
:bomb: Breaking change (it changes the shape of the config).

- [x] https://github.com/ssbc/ssb-meta-feeds/pull/100 should be merged first

### CODE REVIEW

File: template.js [hunk 1]
```diff
@@ -3,117 +3,102 @@
...
 
   hasIndexLeaf() {
-    const indexes = this._find(
-      this._rootNode,
-      (node) => node.feedpurpose === 'indexes'
-    )
-    if (!indexes) return false
-    const leaf = this._find(indexes, (node) => node['$format'] === 'indexed')
-    return !!leaf
+    return this._leafShapes.some((shape) => shape && shape.purpose === 'index')
```
Comment (arj03) [inline template.js:22]: much easier to read 👍

File: test/integration/partial.js [hunk 2]
```diff
@@ -42,6 +45,15 @@ const aliceKeys = u.keysFor('alice')
 const bobKeys = u.keysFor('bob')
 const carolKeys = u.keysFor('carol')
 const davidKeys = u.keysFor('david')
+const aliceMFSeed = Buffer.from(
+  '00000000000000000000000000000000000000000000000000000000000a71ce',
+  'hex'
+)
+const bobMFSeed = Buffer.from(
+  '0000000000000000000000000000000000000000000000000000000000000b0b',
```
Comment (arj03) [inline test/integration/partial.js:53]: :-)

File: test/integration/partial.js [hunk 3]
```diff
@@ -250,20 +264,7 @@ tape('carol acts as an intermediate between alice and bob', async (t) => {
...
-              subfeeds: [
-                {
-                  feedpurpose: 'index',
-                  $format: 'indexed',
-                },
-              ],
-            },
-          ],
-        },
+        1: [{ purpose: 'mygame' }, { purpose: 'index' }],
```
Comment (arj03) [inline test/integration/partial.js:267]: great to see how much simpler this seems to be. The complexity of the feed management into groups is taken care of by the shards. Fingers crossed this works well in practice.

**Comment by arj03** [comment]
The README still says:

```
Requires [ssb-ebt](https://github.com/ssbc/ssb-ebt) version 7.0 or higher
```

shouldn't it now be 9 because of the feedformat stuff?

**Comment by arj03** [comment]
Also still some examples of feedpurpose left in the README:

So that if we write feedpurpose: 'indexes',

File: template.js [hunk 4]
```diff
@@ -3,117 +3,102 @@
...
   }
 
-  matchBranch(branch, mainFeedId) {
-    return this._matchBranch(branch, this._rootNode, mainFeedId)
+  _matchShard(rootMF, shardMF) {
+    if (this._leafShapes.some(isEmptyObject)) return true
+    const wantedShards = this._leafShapes.map((leaf) =>
+      pickShard(rootMF.id, leaf.purpose)
+    )
+    return wantedShards.includes(shardMF.purpose)
```
Comment (arj03) [inline template.js:53]: might be a bit easier to read if this was written like in matchLeaf, so: 

```
this._leafShapes.some((leaf) => {
    if (isEmptyObject(leaf)) return true

    const shardPurpose = pickShard(rootMF.id, leaf.purpose)
    return shardPurpose === shardMF.purpose
})
````

What do you think?

**Comment by arj03** [comment]
Nice, I really like how this is improving the API of this module. Had some minor things, but overall the structure looks good. The dependant PR also looks fine, but I'll let mix to the proper review :-)

**Comment by staltz** [comment]
Thanks for the review! I'm keen to merge them all when I can.

I haven't thought about how group replication would be specified, or how sympathy replication should be, but this PR sets the stage for solving those other problems later, separately. That's what I'll work on next.

**Review by arj03 — APPROVED** [review]
assuming test pass 👍

**Comment by staltz** [comment]
Thanks! I was about to ask if it's okay to merge, good timing. This is going to be a major version update.

**Pull request merged**